Issue/12248 evergreen campaign support

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #12248
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
UI changes to support Evergreen/Endless campaigns on Blaze campaign creation flow. 
- Adds UI changes to Blaze budget screen to switch between endless campaigns and "ending" campaigns. 
- Adds feature flag to hide these changes until the feature is fully ready
- Updates Blaze preview screen to reflect endless campaigns weekly budget
- Updates payment summary screen to reflect `weekly` budget for evergreen campaigns

The changes barely have any logic, is mostly Compose UI updates, hence the absence of Unit tests. 

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

**First test the feature flag works**
After applying this patch none of the changes reflected for `Evergreen` campaigns in the chart below, should be shown.
 
```diff
Subject: [PATCH] Hide anything related to endless campaigns under feature flag
---
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt	(revision 56d1655060bd36a2263729ebf6923cf6b54fdcd8)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt	(date 1723633412447)
@@ -27,11 +27,12 @@
                 PackageUtils.isDebugBuild() || context != null && PackageUtils.isBetaBuild(context)
             }
 
+            ENDLESS_CAMPAIGNS_SUPPORT -> false
+
             WOO_POS,
             WC_SHIPPING_BANNER,
             BETTER_CUSTOMER_SEARCH_M2,
-            ORDER_CREATION_AUTO_TAX_RATE,
-            ENDLESS_CAMPAIGNS_SUPPORT -> PackageUtils.isDebugBuild()
+            ORDER_CREATION_AUTO_TAX_RATE-> PackageUtils.isDebugBuild()
 
             PRODUCT_CREATION_WITH_AI_V2,
             CONNECTIVITY_TOOL,

```


### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
\ | Evergreen | Not evergreen
--- | --- | ---
Budget field | <img src="https://github.com/user-attachments/assets/e3d827d5-51f4-47f3-95e5-339aa4907ba8" width=320 /> | <img src="https://github.com/user-attachments/assets/37d4de2a-62a5-460c-8ef4-f4ca0fc2fb31" width=320 />
Budget screen | <img src="https://github.com/user-attachments/assets/9c07ce97-d2da-4320-a6a1-38a3d39821e5" width=320 /> | <img src="https://github.com/user-attachments/assets/9baa9f62-eb28-4ee2-b514-e6eeb1ba4d5e" width=320 />
Duration screen | <img src="https://github.com/user-attachments/assets/e723baae-df54-41ea-b8cb-23ffc894f718" width=320 /> | <img src="https://github.com/user-attachments/assets/1805061d-76ff-40f2-913a-2ebfeaacb84d" width=320 />
Payment screen | <img src="https://github.com/user-attachments/assets/08d19985-d986-4df6-bf7a-e5dda34e8f14" width=320 /> | <img src="https://github.com/user-attachments/assets/57ad596b-e9f5-4810-9a41-a76df2356589" width=320 />





## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

JorgeMucientes committed:
Add new toggle to select endless (aka Evergreen) campaigns

JorgeMucientes committed:
Fix UI glitches when changing between endless/normal campaigns

JorgeMucientes committed:
Add feature flag for endless campaigns

JorgeMucientes committed:
Set weekly duration when endless campaign is selected

JorgeMucientes committed:
Adjust UI to better support endless campaigns UI

dangermattic commented:
<!--
  0 Errors
  2 Warnings: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-android/milestone/233">20.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

JorgeMucientes committed:
Hide anything related to endless campaigns under feature flag

JorgeMucientes committed:
Show weekly budget for endless campaigns in preview screen

JorgeMucientes committed:
Ceil total budget to avoid decimal values

JorgeMucientes committed:
Ceil total budget to avoid decimal values

JorgeMucientes committed:
Fix detekt indentation issues

JorgeMucientes committed:
Show weekly budget for endless campaigns in payments screen

JorgeMucientes committed:
Restore isEndlessCampaign value from navargs

JorgeMucientes committed:
Fix unit tests

JorgeMucientes committed:
Add some minimal padding to blaze preview campaign detail items

JorgeMucientes committed:
Add some minimal padding to blaze preview campaign detail items

JorgeMucientes committed:
Add some minimal padding to blaze preview campaign detail items

JorgeMucientes committed:
Fix color of switch when unchecked in dark mode

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/BlazeRepository.kt:
```diff
@@ -141,7 +142,8 @@ class BlazeRepository @Inject constructor(
             currencyCode = BLAZE_DEFAULT_CURRENCY_CODE,
             durationInDays = DEFAULT_CAMPAIGN_DURATION,
-            startDate = Date().apply { time += 1.days.inWholeMilliseconds }, // By default start tomorrow
+            startDate = Date().apply { time += 1.days.inWholeMilliseconds }, // By default start tomorrow,
```
np: There's an extra comma at the end of the comment. I wonder if the auto formatting added it.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetScreen.kt:
```diff
@@ -348,38 +361,65 @@ private fun EditDurationBottomSheet(
-                inactiveTrackColor = colorResource(id = color.divider_color)
-            )
-        )
+        AnimatedVisibility(isEndlessCampaign.not()) {
```
The animation looks nice! 👍🏻

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt:
```diff
@@ -206,9 +212,20 @@ class BlazeCampaignBudgetViewModel @Inject constructor(
+        isEndlessCampaign: Boolean
+    ): String {
+        return if (isEndlessCampaign) {
+            resourceProvider.getString(
```
Using `ResourceProvider` in view models has side effects. There is an open issue, #12116 to remove all `ResourceProvider`s. 
In this case, removing the use of `ResourceProvider` for `campaignDurationDisplayText` is not straightforward. If you can find a simple way to avoid using ResourceProvider here, feel free to do so. Otherwise, you can skip it for now, and we can address it in #12116.

JorgeMucientes replied:
Good point; I was unaware that was agreed upon. However, as mentioned in the discussion about the side effects of using `resourceProvider` in viewModels, the issues are corner case and minor (I'm referring to the language not getting updated due to retained app context). 
In any case, I've got rid of the use of resource provider in this particular case.

irfano replied:
It's better now. 🙏🏻 Thanks!

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/payment/BlazeCampaignPaymentSummaryViewModel.kt:
```diff
@@ -150,8 +149,23 @@ class BlazeCampaignPaymentSummaryViewModel @Inject constructor(
+            currencyCode = navArgs.campaignDetails.budget.currencyCode
+        )
+        return when {
+            navArgs.campaignDetails.budget.isEndlessCampaign -> resourceProvider.getString(
```
It seems easier to avoid using `ResourceProvider` here. I’d like to hear your thoughts on this.

irfano commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetViewModel.kt:
```diff
@@ -206,9 +212,20 @@ class BlazeCampaignBudgetViewModel @Inject constructor(
+        return if (isEndlessCampaign) {
+            resourceProvider.getString(
+                R.string.blaze_campaign_budget_duration_section_endless_campaign_value,
+                Date(startDateMillis).formatToMMMdd()
```
Using hardcoded patterns for date formatting causes localization issues. However, this already exists throughout the entire app. I have opened an [issue](https://github.com/woocommerce/woocommerce-android/issues/12401) for it.

JorgeMucientes replied:
Yes, good catch @irfano . This is a long overdue issue in the app that we haven't prioritized yet. 

For this particular case, I took the chance to partially fix the problem. I fixed the function `formatToMMMddYYYY` https://github.com/woocommerce/woocommerce-android/pull/12324/commits/9c470750d43741ff037e7469cd62f5e4357b4ae9 to apply that same hardcoded format but in a localized way. I fixed this particular case because it was straightforward to do, as using `SimpleDateFormat.MEDIUM` provides the equivalent format to `MMMddYYYY` in a localized style.
 
However, for many other formats we are directly using like `formatToMMMdd` there isn't an equivalent format provided by `DateFormat` constants. So fixing that won't be straightforward. I'm not even sure that'd be possible, because some date formats we are hardcoding are very specific.

irfano replied:
Thanks for fixing this one. 👍🏻

irfano reviewed (approved):
I tested the added feature, and it works as expected. Good job! I added some questions and minor comments. I'm approving it but not merging, in case you'd like to make any changes.

JorgeMucientes committed:
Fix formatToMMMddYYYY to apply the pattern in a localized manner

JorgeMucientes committed:
Remove resource provider usages from BlazeCampaignBudgetViewModel

JorgeMucientes committed:
Format start date for endless campaigns consistently

JorgeMucientes committed:
Merge branch 'refs/heads/trunk' into issue/12248-evergreen-campaign-support

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/blaze/creation/budget/BlazeCampaignBudgetScreen.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/util/FeatureFlag.kt

JorgeMucientes committed:
Remove extra comma from comment

JorgeMucientes commented:
Thanks for the detailed feedback @irfano (and sorry for not getting back to you sooner. 

I've addressed your suggestion (some of the partially, see my other comment). Let me know what you think about the applied changes.

irfano reviewed (commented):
Thanks for addressing my comments. We're good to merge after fixing the failing unit tests.

JorgeMucientes committed:
Replace UK with default for unit tests

The expected date format for UK Locale was incorrect, UK locale will place the month name before the day and will make the tests fail when we use a Locale based date formatting

JorgeMucientes committed:
Merge branch 'trunk' into issue/12248-evergreen-campaign-support
